src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -169,42 +171,105 @@ void NnetDecodableBase::GetOutputForFrame(int32 subsampled_frame,
+    int32 num_ivectors_left = std::ceil(1.0 * left_context / interval);
+    int32 num_ivectors_right = (num_output_frames + right_context - 1) /
+                               interval + 1;
+    ivector->Resize(num_ivectors_left + num_ivectors_right,
+                    online_ivector_feats_->NumCols(), kUndefined);
+    int32 frame_to_search_left = output_t_start - 1 -
+                                 std::min(interval, left_context) / 2;
+    int32 ivector_frame = frame_to_search_left / online_ivector_period_;
+    KALDI_ASSERT(ivector_frame >= 0);
+    for (int32 n = 0; n < num_ivectors_left; n++) {
+      if (ivector_frame < 0) {
+        int32 margin = -ivector_frame;
+        if (margin * online_ivector_period_ > 50) {
+          // Half a second seems like too long to be explainable as edge effects.
+          KALDI_ERR << "Could not get iVector for frame "
+                    << frame_to_search_left
+                    << ", only available till frame "
+                    << "0 * ivector-period=" << online_ivector_period_
+                    << " (mismatched --ivector-period?)";
+        }
+        ivector_frame = 0;
+      }
+      int32 row = num_ivectors_left - 1 - n;
+      ivector->Row(row).CopyFromVec(online_ivector_feats_->Row(ivector_frame));
+      frame_to_search_left -= interval;
+      ivector_frame = frame_to_search_left / online_ivector_period_;
+    }
+    int32 frame_to_search_right = output_t_start +
+                                  std::min(interval, num_output_frames) / 2;
+    ivector_frame = frame_to_search_right / online_ivector_period_;
+    KALDI_ASSERT(ivector_frame >= 0);
+    for (int32 n = 0; n < num_ivectors_right; n++) {
+      if (ivector_frame >= online_ivector_feats_->NumRows()) {
+        int32 margin = ivector_frame - (online_ivector_feats_->NumRows() - 1);
+        if (margin * online_ivector_period_ > 50) {
+          // Half a second seems like too long to be explainable as edge effects.
+          KALDI_ERR << "Could not get iVector for frame "
+                    << frame_to_search_right
+                    << ", only available till frame "
+                    << online_ivector_feats_->NumRows()
+                    << " * ivector-period=" << online_ivector_period_
+                    << " (mismatched --ivector-period?)";
+        }
+        ivector_frame = online_ivector_feats_->NumRows() - 1;
+      }
+      int32 row = num_ivectors_left + n;
+      ivector->Row(row).CopyFromVec(online_ivector_feats_->Row(ivector_frame));
+      frame_to_search_right += interval;
+      ivector_frame = frame_to_search_right / online_ivector_period_;
     }
-    ivector_frame = online_ivector_feats_->NumRows() - 1;
   }
-  *ivector = online_ivector_feats_->Row(ivector_frame);
 }
 
 
 void NnetDecodableBase::DoNnetComputation(
     int32 input_t_start,
     const MatrixBase<BaseFloat> &input_feats,
-    const VectorBase<BaseFloat> &ivector,
+    const MatrixBase<BaseFloat> &ivector,
```

Comment:
this interface isn't very clear after the change.  why is it still called ivector?
And there is now no indication of which frame each ivector belongs to.
The code for GetCurrentIvector is very odd too.  Since we are now assigning iVectors to specific frames and not to the chunk, the structure of the code should change.   Also the way you treat the left and right contexts separately doesn't seem right to me.  And the interface of DoNnetComputation isn't right because now that the ivector is a matrix, the ivector_period should be supplied to help interpretation of the matrix.
I'm wondering whether it's best if I do the decodable part myself.